crates/common/tedge_config/src/tedge_config_cli/figment.rs

```diff
@@ -0,0 +1,263 @@
+use std::borrow::Cow;
+use std::fmt::Display;
+use std::path::Path;
+use std::path::PathBuf;
+
+use figment::providers::Format;
+use figment::providers::Toml;
+use figment::Figment;
+use figment::Metadata;
+use serde::de::DeserializeOwned;
+
+use crate::TEdgeConfigError;
+
+/// Extract the configuration data from the provided TOML path and `TEDGE_` prefixed environment variables
+pub fn extract_data<T: DeserializeOwned>(path: impl AsRef<Path>) -> Result<T, TEdgeConfigError> {
+    let env = TEdgeEnv::default();
+    let figment = Figment::new().merge(Toml::file(path)).merge(env.provider());
+
+    let data = extract_exact(&figment, &env);
+
+    for warning in unused_value_warnings::<T>(&figment, &env)
+        .ok()
+        .unwrap_or_default()
+    {
+        tracing::warn!("{warning}");
+    }
+
+    data
+}
+
+fn unused_value_warnings<T: DeserializeOwned>(
+    figment: &Figment,
+    env: &TEdgeEnv,
+) -> Result<Vec<String>, TEdgeConfigError> {
+    let mut warnings = Vec::new();
+
+    let value = extract_exact::<toml::Value>(figment, env)?;
+
+    // Serializing and deserializing again is the only way I could find to use serde_ignored
```

Comment:
I guess this is because you want to find values that are provided either by the env or the config file, but actually not translated into some TEdgeConfig? Am I correct? If so this is not chocking to serialize/deserialize the merge of the two sources; as serde_ignored work on deserialization.

Reply:
Yeah. It would be nice if figment exposed a deserializer that we could pass in and get warnings for when we do the main deserialization, but this does mean we don't need a macro to collect the fields and another macro to check whether any of them are populated.

I might have a play with patch dependencies locally to see if exposing a deserializer is trivially possible and submit a feature request to figment if that's the case.